### PR TITLE
[algo, doc] feat: trust region sequence masking - (1) k3 KL avg and (2) veto for max criterion

### PR DESCRIPTION
**Pull request by szrlee**
```bibtex
@article{li2025trustregion,
  title   = {Trust Region Masking for Long-Horizon LLM Reinforcement Learning},
  author  = {Li, Yingru and Liu, Jiacai and Xu, Jiawei and Tong, Yuxuan and Li, Ziniu and Wang, Baoxiang},
  journal = {arXiv preprint arXiv:2512.23075},
  year    = {2025},
  url     = {https://arxiv.org/abs/2512.23075}
}
```

Link: https://richardli.xyz/post/trust-region-masking/

Summary
- Add K1/K3 KL divergence estimators and modify token veto mechanism for rollout correction.

New Features
- Introduce sequence-level `k1` ($|\mathbb{E}[\log(ratio)]|$, divergence-based) and `k3` ($\mathbb{E}[r - \log(r) - 1]$, small-KL stable) rollout sequence masking modes.
- Expand factory presets across K1, K3, geometric, and token TIS modes (e.g., `decoupled_k1_rs`, `bypass_pg_k3_rs`).
- Rename internal helpers (`rollout_is_weights` → `rs_statistic`) and introduce a consistent `decoupled_` prefix for decoupled presets.
- Refine the token veto mechanism: use $|\log(r)|$ to flag catastrophic tokens and drop any sequence that contains them.

Semantic Clarifications
- Clarify divergence vs geometric semantics: K1 targets $|\mathbb{E}[\log(r)]|$ (ideal 0.0, threshold ~0.001) while geometric modes target $\exp(\mathbb{E}[\log(r)])$ (ideal 1.0, threshold ~1.001).
- Document that K3 uses `log_ratio_safe` for stability and highlight threshold guidance across modes.

Documentation
- Add mathematical formulations for K1 and K3, including estimator × operating mode comparison tables.
- Provide a comprehensive preset reference for all 20+ factory methods and clarify naming/threshold semantics.

Recommendations
- Use `k3` mode for average-based filtering since K3 is non-negative and unbiased estimator.
- Use token veto mechanism for max-based filtering since |log(ratio)| is a symmetric detector and both $ratio \gg 1$  and $ratio \ll 1$ indicate large divergence.

Test Plan
- Unit tests covering new `k1` and `k3` rollout rejection sampling modes.
- Unit tests for token veto mechanism using |log(r)| to reject sequences with catastrophic tokens.
- Run existing rollout correction test suites to confirm backward compatibility.

----

## Solution: Trust Region Masking (TRM)

We propose **Trust Region Masking (TRM)**, which excludes entire sequences from gradient computation if any token violates the trust region.

### Why Sequence Masking Works
By masking at the sequence level, we ensure that:
1. The remaining samples come from policies that are provably close
2. The gradient estimator remains unbiased over the unmasked samples
3. The trust region bound applies to all included sequences

### Masking Criterion
A sequence is masked if:
$$\max_{t} D_{\mathrm{KL}}(\pi_{\mathrm{roll}}(\cdot|c_t) \| \pi_{\theta}(\cdot|c_t)) > \epsilon$$

where $\epsilon$ is the trust region threshold.

### Exact Computation
The rigorous guarantee requires **exact KL computation with stored logits** from the rollout policy.

### Sample-Based Approximation

In practice, storing full logits may be expensive. The paper proposes sample-based approximations using importance ratios $\rho_t = \frac{\pi_{\theta}(y_t|c_t)}{\pi_{\mathrm{roll}}(y_t|c_t)}$:

#### The $k_3$ Estimator (for average-based filtering)
$$k_3(\rho) = \rho - 1 - \log \rho$$

| $\rho$ | $k_1 = -\log \rho$ | $k_3 = \rho - 1 - \log \rho$ |
|--------|-------------------|------------------------------|
| 0.5 | 0.69 | 0.19 |
| 1.0 | 0 | 0 |
| 2.0 | −0.69 | 0.31 |
| 10 | −2.30 | 6.70 |
| 100 | −4.61 | 94.4 |

**Properties of $k_3$:**
- **Non-negative**: $\rho - 1 - \log \rho \geq 0$ for all $\rho > 0$
- **Unbiased**: $\mathbb{E}_{y \sim \pi_{\mathrm{roll}}}[k_3(\rho)] = D_{\mathrm{KL}}$
- Ideal for average-based filtering since $(1/T)\sum_t k_3(\rho_t)$ converges to true average KL

#### The $|\log \rho|$ Estimator (for max-based filtering)
For the max criterion, we need a **symmetric** detector since both $\rho \gg 1$ and $\rho \ll 1$ indicate large divergence:
$$|\log(100)| = |\log(0.01)| = 4.6$$

In contrast, $k_3$ is asymmetric: $k_3(100) = 94.4$ but $k_3(0.01) = 3.6$ (26× difference).

#### Caveat
Neither sample-based method provides a rigorous *bound* on $D_{\mathrm{KL}}^{\mathrm{tok,max}}$ — both are approximate detectors based on single samples per context.

### CODE REVIEW

File: verl/trainer/config/algorithm.py [hunk 1]
```diff
@@ -90,9 +90,17 @@ class RolloutCorrectionConfig(BaseConfig):
...
             - None: No rejection sampling
-            - "token": Reject individual tokens with outlier ratios
-            - "sequence": Reject entire sequences with outlier ratios
-            - "geometric": Geometric mean aggregation (threshold: 1.0002-1.001)
+            - "token": Reject individual tokens with outlier ratios (ideal=1.0)
+            - "sequence": Reject entire sequences with outlier ratios (ideal=1.0)
+            - "geometric": Geometric mean IS ratio exp(E[log(r)]) (ideal=1.0)
+              Length-normalized ratio, uses [lower, upper] threshold bounds.
+              (typical: 1.0002-1.001)
+            - "k1": K1 KL divergence at sequence level: |E[log(r)]|
```
Comment (tongyx361) [inline verl/trainer/config/algorithm.py]: iiuc, the $k_1$ estimator is defined as $- \log r$?
This option might be better called `"abs_mean_k1"`, also stressing that the absolute value is applied after the mean?
Should we also separate the true expectation $\mathbb{E}$ and the MC estimation $\hat{\mathbb{E}}$, like using "1/N * estimator" instead?

File: verl/trainer/config/algorithm.py [hunk 2]
```diff
@@ -107,9 +115,10 @@ class RolloutCorrectionConfig(BaseConfig):
 
         rollout_token_veto_threshold (Optional[float]): Per-token veto for catastrophic outliers.
             - Checks unclamped per-token ratios before safety bounds
-            - If ANY token has ratio < threshold, entire sequence is rejected
+            - If ANY token has abs(π_train/π_rollout) > threshold, the entire sequence is rejected
```
Comment (tongyx361) [inline verl/trainer/config/algorithm.py]: Isn't the "π_train/π_rollout" always positive?

File: verl/trainer/ppo/rollout_corr_helper.py [hunk 3]
```diff
@@ -102,71 +113,126 @@ def compute_rollout_rejection_mask(
...
+        log_ratio_for_metrics = log_ratio_mean  # Store log-space for metrics
 
         log_ratio_mean_safe: torch.Tensor = torch.clamp(log_ratio_mean, min=-SAFETY_BOUND, max=SAFETY_BOUND)
-        rollout_is_weights = torch.exp(log_ratio_mean_safe).expand_as(log_ratio)
+        rs_statistic = torch.exp(log_ratio_mean_safe).expand_as(log_ratio)  # Geometric mean ratio
+
+    elif rollout_rs == "k1":
+        # K1 divergence at sequence level: |E[log(r)]|
+        # Use absolute value for symmetric divergence (always >= 0, ideal = 0)
+        # This equals KL(π_rollout || π_train) in expectation (the reverse KL)
```
Comment (tongyx361) [inline verl/trainer/ppo/rollout_corr_helper.py]: This doesn't equal KL(π_rollout || π_train) in expectation since there is an absolute value?

File: verl/trainer/ppo/rollout_corr_helper.py [hunk 4]
```diff
@@ -102,71 +113,126 @@ def compute_rollout_rejection_mask(
...
+        log_ratio_mean: torch.Tensor = verl_F.masked_mean(log_ratio, response_mask, axis=-1).unsqueeze(
```
Comment (tongyx361) [inline verl/trainer/ppo/rollout_corr_helper.py]: To derive the sequence-level log probability ratio from token-level ones, shouldn't we sum instead of mean?

File: verl/trainer/ppo/rollout_corr_helper.py [hunk 5]
```diff
@@ -102,71 +113,126 @@ def compute_rollout_rejection_mask(
...
+        # K3 divergence at sequence level: E[r - log(r) - 1]
+        # where r = π_train/π_rollout = exp(log_ratio)
+        # K3 >= 0 per token (equals 0 when r=1), equals KL(π_rollout || π_train) in expectation
+        # More stable than K1 because each token contribution is non-negative
+        log_ratio_safe: torch.Tensor = torch.clamp(log_ratio, min=-SAFETY_BOUND, max=SAFETY_BOUND)
+        r = torch.exp(log_ratio_safe)  # Token-level ratio: r = exp(log_ratio_safe)
+        k3_token = r - log_ratio_safe - 1  # K3 = r - log(r) - 1, using log(r) = log_ratio_safe
+
+        # Sequence-level K3: mean of token K3 values
+        k3_seq: torch.Tensor = verl_F.masked_mean(k3_token, response_mask, axis=-1).unsqueeze(-1)
```
Comment (tongyx361) [inline verl/trainer/ppo/rollout_corr_helper.py]: 1. To derive the sequence-level KL from token-level ones, shouldn't we also sum instead of mean?
2. Should we derive the sequence-level $k_3$ estimator by first calculating the sequence-level probability ratio? Both (a) estimating with the sequence-level probability ratio and (b) summing up token-level estimations should be reasonable, but now "k1" is using (a) while "k3" is using (b). Should we be at least consistent? 

Please correct me if I'm wrong.

Comment (szrlee) [reply verl/trainer/ppo/rollout_corr_helper.py]: cc @jiawei415

File: verl/trainer/config/algorithm.py [hunk 6]
```diff
@@ -88,29 +88,35 @@ class RolloutCorrectionConfig(BaseConfig):
...
+            - Reduces variance by ensuring average weight is 1.0 per batch
+            - Only affects IS weight values, not rejection sampling
+            Default: False (no batch normalization)
+
+        rollout_rs (Optional[str]): Rejection sampling aggregation modes.
+            Accepts a comma-delimited list (duplicates removed) of canonical options implemented in
+            ``rollout_corr_helper``:
+            - "token_k1": Token-level ``-log r`` bounds (ratio thresholds supplied via
+              ``rollout_rs_threshold`` as ``lower_upper``)
+            - "token_k2": Token-level χ²/2 estimator ``0.5 * (log r)^2`` (upper bound only)
```
Comment (tongyx361) [inline verl/trainer/config/algorithm.py]: Why is this "χ²/2"?

File: verl/trainer/config/algorithm.py [hunk 7]
```diff
@@ -88,29 +88,35 @@ class RolloutCorrectionConfig(BaseConfig):
...
+            - "token_k3": Token-level third-order KL estimator ``exp(log r) - 1 - log r`` (upper bound only)
```
Comment (tongyx361) [inline verl/trainer/config/algorithm.py]: What does third-order mean?

**Review by tongyx361 — COMMENTED** [review]
Why are the names so weird? Maybe I can help refactor.

File: verl/trainer/config/algorithm.py (hunk 6 above)
Comment (jiawei415) [reply verl/trainer/config/algorithm.py]: used the wrong notion for k2 KL estimator

File: verl/trainer/config/algorithm.py (hunk 7 above)
Comment (jiawei415) [reply verl/trainer/config/algorithm.py]: should be k3 KL estimator

**Comment by LeonEricsson** [comment]
@szrlee @tongyx361 
I note the defaults for Seq-MIS and Geo-RS use the k1 estimator (`seq_sum_k1` and `seq_mean_k1`). Is k3 not the preferred choice here given its lower variance and strict non-negativity?
